frameworklauncher/src/main/java/com/microsoft/frameworklauncher/common/model/ResourceDescriptor.java

```diff
@@ -130,82 +190,156 @@ public static ResourceDescriptor fromResource(Resource res) throws Exception {
+        setRangesList.invoke(obj, hadoopValueRangeList);
+
+        Method setPort = clazz.getMethod("setPorts", hadoopValueRangesClass);
+        setPort.invoke(res, obj);
+
+      } catch (NoSuchMethodException e) {
+        LOGGER.logDebug(e, "Ignore: Failed to get Ports information, YARN library doesn't support Port");
+      } catch (IllegalAccessException e) {
+        LOGGER.logError(e, "Ignore: Failed to get Ports information, illegal access function");
+      } catch (ClassNotFoundException e) {
+        LOGGER.logDebug(e, "Ignore: Failed to get the class Name");
+      }
+    }
+    LOGGER.logDebug("Put LocalResource " + this.toString() + " to hadoop resource: " + res);
+    return res;
   }
 
   // Maybe underestimate if any GpuAttribute == 0
   public static ResourceDescriptor subtract(ResourceDescriptor lhs, ResourceDescriptor rhs) {
-    ResourceDescriptor ret = new ResourceDescriptor();
-    ret.setMemoryMB(lhs.getMemoryMB() - rhs.getMemoryMB());
-    ret.setCpuNumber(lhs.getCpuNumber() - rhs.getCpuNumber());
-    ret.setGpuAttribute(lhs.getGpuAttribute() & (~(rhs.getGpuAttribute())));
-    if (lhs.getGpuAttribute() != 0 && rhs.getGpuAttribute() != 0) {
-      ret.setGpuNumber(Long.bitCount(ret.getGpuAttribute()));
-    } else {
-      ret.setGpuNumber(lhs.getGpuNumber() - rhs.getGpuNumber());
-    }
+    ResourceDescriptor ret = YamlUtils.deepCopy(lhs, ResourceDescriptor.class);
+    ResourceDescriptor.subtractFrom(ret, rhs);
     return ret;
   }
 
-  // Maybe overestimate if any GpuAttribute == 0
+  public static ResourceDescriptor subtractFrom(ResourceDescriptor lhs, ResourceDescriptor rhs) {
+    lhs.setMemoryMB(lhs.getMemoryMB() - rhs.getMemoryMB());
+    lhs.setCpuNumber(lhs.getCpuNumber() - rhs.getCpuNumber());
+    lhs.setGpuAttribute(lhs.getGpuAttribute() & (~(rhs.getGpuAttribute())));
+    lhs.setGpuNumber(Long.bitCount(lhs.getGpuAttribute()));
+    lhs.setPortRanges(ValueRangeUtils.subtractRange(lhs.getPortRanges(), rhs.getPortRanges()));
+    lhs.setPortNumber(ValueRangeUtils.getValueNumber(lhs.getPortRanges()));
+    return lhs;
+  }
+
   public static ResourceDescriptor add(ResourceDescriptor lhs, ResourceDescriptor rhs) {
-    ResourceDescriptor ret = new ResourceDescriptor();
-    ret.setMemoryMB(lhs.getMemoryMB() + rhs.getMemoryMB());
-    ret.setCpuNumber(lhs.getCpuNumber() + rhs.getCpuNumber());
-    ret.setGpuAttribute(lhs.getGpuAttribute() | rhs.getGpuAttribute());
-    if (lhs.getGpuAttribute() != 0 && rhs.getGpuAttribute() != 0) {
-      ret.setGpuNumber(Long.bitCount(ret.getGpuAttribute()));
-    } else {
-      ret.setGpuNumber(lhs.getGpuNumber() + rhs.getGpuNumber());
-    }
+    ResourceDescriptor ret = YamlUtils.deepCopy(lhs, ResourceDescriptor.class);
+    ResourceDescriptor.addTo(ret, rhs);
     return ret;
   }
 
+  // Maybe overestimate if any GpuAttribute == 0
+  public static void addTo(ResourceDescriptor lhs, ResourceDescriptor rhs) {
```

Comment:
>addTo [](start = 21, length = 5)

Can we remove this function? We do not want to change input always.

Reply:
No need deepcopy, please revert to the original way for this method,

---
In reply to: [174759681](https://github.com/Microsoft/pai/pull/318#discussion_r174759681) [](ancestors = 174759681)

Reply:
please revert to the original way for this method,

---
In reply to: [175414255](https://github.com/Microsoft/pai/pull/318#discussion_r175414255) [](ancestors = 175414255,174759681)